packages/core/src/stream-loading/state-manipulator.ts

```diff
@@ -48,12 +49,24 @@ export class StateManipulator {
     private readonly logSyncer: LogSyncer
   ) {}
 
+  /**
+   * Applies log to state and returns a new state.  Most callers should set 'cloneInput' to true
+   * so that the original StreamState is not modified.  Some codepaths may set 'cloneInput' to false
+   * if they are confident that the StreamState has already been cloned once within the
+   * StateManipulator and so the input state that was given to the StateManipulator will not be
+   * modified.
+   */
   async _applyLog<T extends Stream>(
     handler: StreamHandler<T>,
     state: StreamState | null,
     log: AppliableStreamLog,
-    throwOnInvalidCommit: boolean
+    throwOnInvalidCommit: boolean,
+    cloneInput: boolean
```

Comment:
That seems like a potential source of issues if relying of the caller to know if an object is safe to mutate or not. Could we instead make objects that get passed around immutable so the functions using them can't accidentally mutate them? I think that would also be more efficient than always cloning "just in case".

Reply:
Is there actually a way to make objects immutable in js?

Reply:
There are full-feature libraries such as https://immutable-js.com/ that provide APIs to manipulate immutable structures, but if we don't want to leverage a dedicated library, at least we could use [`Object.freeze()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/freeze) to leverage the guaranties that can be provided by the JS engine.
Considering our usage of cloning objects, I think we could instead freeze them on creation and use TS' [`Readonly` type](https://www.typescriptlang.org/docs/handbook/utility-types.html#readonlytype) in arguments when passing these objects along to ensure the functions consuming them don't try to mutate them.

Reply:
Ooh okay, I kind of like using the Readonly type, though that will mean changing the signature of all the StreamTypeHandlers.  I'm going to leave this PR as-is for now, but I have https://linear.app/3boxlabs/issue/CDB-2767/figure-out-best-way-to-ensure-streamtype-handlers-dont-mutate-input to follow-up on this.